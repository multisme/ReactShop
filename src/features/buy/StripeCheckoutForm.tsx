import React, { useState } from "react";
import { Link, useHistory } from "react-router-dom";

import { CardElement, useStripe, useElements } from "@stripe/react-stripe-js";
import { loadStripe, PaymentMethod, StripeError } from "@stripe/stripe-js";

const CARD_OPTIONS = {
  style: {
    base: {
      fontSize: "16px",
    },
    invalid: {
      iconColor: "#ffc7ee",
      color: "#ffc7ee",
    },
  },
};

interface StripeCheckoutFormState {
  total: number;
}

const StripeCheckoutForm = ({ total }: StripeCheckoutFormState) => {
  const stripe = useStripe();
  const elements = useElements();
  const history = useHistory();

  const handleSubmit = async (event: any) => {
    event.preventDefault();

    if (!stripe || !elements) {
      // Stripe.js has not loaded yet. Make sure to disable
      // form submission until Stripe.js has loaded.
      return;
    }

    const card = elements.getElement("card");
    if (card == null) {
      return;
    }

    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: "card",
      card: card,
    });
    if (error) {
      history.push("/error");
    }
    if (paymentMethod) {
      history.push("/thank");
    }
  };
  return (
    <form onSubmit={handleSubmit}>
      <CardElement options={CARD_OPTIONS} className={"stripePaiment"} />
      <button type="submit" disabled={!stripe}>
        Pay {total}â‚¬
      </button>
      <Link to="/home">
        <button>CANCEL</button>
      </Link>
    </form>
  );
};

export default StripeCheckoutForm;
