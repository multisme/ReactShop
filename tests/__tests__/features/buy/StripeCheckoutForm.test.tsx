import { mount } from "enzyme";
import { Elements } from "@stripe/react-stripe-js";

import StripeCheckoutForm from "features/buy/StripeCheckoutForm";
import { loadStripe } from "@stripe/stripe-js";
import { MemoryRouter } from "react-router-dom";

describe("StripeCheckoutForm", () => {
  const stripePromise = loadStripe(
    "pk_test_esdk3EP8hqcjKThBW7kK36Vv00Ah7Cskxb"
  );
  let stripeMock: any;
  beforeEach(() => {
    const elementMock = {
      mount: jest.fn(),
      destroy: jest.fn(),
      on: jest.fn(),
      update: jest.fn(),
    };
    const elementsMock = {
      create: jest.fn().mockReturnValue(elementMock),
    };
    stripeMock = {
      elements: jest.fn().mockReturnValue(elementsMock),
      createToken: jest.fn(),
      createSource: jest.fn(),
      createPaymentMethod: jest.fn(),
      handleCardPayment: jest.fn(),
      handleCardSetup: jest.fn(),
    };

    window.Stripe = jest.fn().mockReturnValue(stripeMock);
  });
  it("checks that the total is correct", () => {
    const component = mount(
      <MemoryRouter>
        <Elements stripe={stripePromise}>
          <StripeCheckoutForm total={42} />
        </Elements>
      </MemoryRouter>
    );
    const button = component.find('button[type="submit"]');
    expect(button.text()).toEqual("Pay 42â‚¬");
    expect(component).toMatchSnapshot();
  });
  it("goes to ThanksPage page if paiement succeed", () => {
    stripeMock.createPaymentMethod.mockReturnValue({
      error: null,
      paymentMethod: jest.fn(),
    });
    const component = mount(
      <MemoryRouter>
        <Elements stripe={stripePromise}>
          <StripeCheckoutForm total={42} />
        </Elements>
      </MemoryRouter>
    );
    const button = component.find('button[type="submit"]').simulate("click");
    expect(stripeMock.createPaymentMethod).toHaveBeenCalled();
  });
  it("goes to ErrorPaiement page if paiement succeed", () => {
    const component = mount(
      <MemoryRouter>
        <Elements stripe={stripePromise}>
          <StripeCheckoutForm total={42} />
        </Elements>
      </MemoryRouter>
    );
    const button = component.find('button[type="submit"]').simulate("click");
  });
});
