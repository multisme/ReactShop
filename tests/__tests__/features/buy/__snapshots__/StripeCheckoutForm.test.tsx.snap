// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StripeCheckoutForm checks that the total is correct 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "vcr6jq",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "vcr6jq",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Elements
      stripe={Promise {}}
    >
      <StripeCheckoutForm
        total={42}
      >
        <form
          onSubmit={[Function]}
        >
          <CardElement
            className="stripePaiment"
            options={
              Object {
                "style": Object {
                  "base": Object {
                    "fontSize": "16px",
                  },
                  "invalid": Object {
                    "color": "#ffc7ee",
                    "iconColor": "#ffc7ee",
                  },
                },
              }
            }
          >
            <div
              className="stripePaiment"
            />
          </CardElement>
          <button
            disabled={true}
            type="submit"
          >
            Pay 
            42
            â‚¬
          </button>
          <Link
            to="/home"
          >
            <LinkAnchor
              href="/home"
              navigate={[Function]}
            >
              <a
                href="/home"
                onClick={[Function]}
              >
                <button>
                  CANCEL
                </button>
              </a>
            </LinkAnchor>
          </Link>
        </form>
      </StripeCheckoutForm>
    </Elements>
  </Router>
</MemoryRouter>
`;
